sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/kafka/BeamKafkaProtoTable.java

```diff
@@ -0,0 +1,140 @@
+    return new ProtoRecorderEncoder<>(protoClass);
+  }
+
+  /** A PTransform to convert {@code KV<byte[], byte[]>} to {@link Row}. */
+  private static class ProtoRecorderDecoder<ProtoT extends Message>
+      extends PTransform<PCollection<KV<byte[], byte[]>>, PCollection<Row>> {
+    private final Schema schema;
+    private final ProtoCoder<ProtoT> protoCoder;
+    private final SerializableFunction<ProtoT, Row> toRowFunction;
+
+    ProtoRecorderDecoder(Schema schema, Class<ProtoT> clazz) {
+      this.schema = schema;
+      this.protoCoder = ProtoCoder.of(clazz);
+      this.toRowFunction = new ProtoMessageSchema().toRowFunction(TypeDescriptor.of(clazz));
+    }
+
+    @Override
+    public PCollection<Row> expand(PCollection<KV<byte[], byte[]>> input) {
+      return input
+          .apply(
+              "decodeProtoRecord",
+              ParDo.of(
+                  new DoFn<KV<byte[], byte[]>, Row>() {
+                    @ProcessElement
+                    public void processElement(ProcessContext c) {
+                      Row decodedRow = decodeBytesToRow(c.element().getValue());
+                      c.output(decodedRow);
+                    }
+                  }))
+          .setRowSchema(schema);
+    }
+
+    private Row decodeBytesToRow(byte[] bytes) {
+      try {
+        InputStream inputStream = new ByteArrayInputStream(bytes);
+        ProtoT message = protoCoder.decode(inputStream);
+        return toRowFunction.apply(message);
+      } catch (IOException e) {
+        throw new IllegalArgumentException("Could not decode row from proto payload.", e);
+      }
+    }
+  }
+
+  /** A PTransform to convert {@link Row} to {@code KV<byte[], byte[]>}. */
+  private static class ProtoRecorderEncoder<ProtoT extends Message>
+      extends PTransform<PCollection<Row>, PCollection<KV<byte[], byte[]>>> {
+    private final SerializableFunction<Row, ProtoT> toProtoFunction;
+    private final ProtoCoder<ProtoT> protoCoder;
+    private final Class<ProtoT> clazz;
+
+    public ProtoRecorderEncoder(Class<ProtoT> clazz) {
+      this.protoCoder = ProtoCoder.of(clazz);
+      this.toProtoFunction = new ProtoMessageSchema().fromRowFunction(TypeDescriptor.of(clazz));
+      this.clazz = clazz;
+    }
+
+    @Override
+    public PCollection<KV<byte[], byte[]>> expand(PCollection<Row> input) {
+      return input.apply(
+          "encodeProtoRecord",
+          ParDo.of(
+              new DoFn<Row, KV<byte[], byte[]>>() {
+                @ProcessElement
+                public void processElement(ProcessContext c) {
+                  Row row = c.element();
+                  c.output(KV.of(new byte[] {}, encodeRowToProtoBytes(row)));
+                }
+              }));
+    }
+
+    byte[] encodeRowToProtoBytes(Row row) {
+      ByteArrayOutputStream outputStream = new ByteArrayOutputStream();
+      try {
+        Message message = toProtoFunction.apply(row);
```

Comment:
There's the possibility for some subtle errors here because SQL's concept of the schema could have the fields in a different order than the order in the protobuf descriptor, lets make sure to test that case, and consider adding a step to "convert" between the protobuf-generated schema and the SQL schema when encoding and decoding.

Reply:
Done.